### PR TITLE
generate: support nested records.

### LINKED ISSUE #128: Pseudo-"record" support for Rust.
Post-https://github.com/LykenSol/grammer/pull/8, it might be worth exploring record emulation in Rust.

One idea would be to generate a module with structs, e.g. for the example in that PR:
```rust
mod _rec {
    pub struct b__<b> {
        b: b,
    }
    pub struct c__d__<c, d> {
        c: c,
        d: d,
    }
}
```
Then they could be used wherever those field names are needed:
```rust
struct Foo {
    a: _rec::b__<_rec::c__d__<Handle, Handle>>,
}
```
And that way regular field accesses (e.g. `foo.a.b.c` and `foo.a.b.d`) would work.

### PR DESCRIPTION
**Pull request by eddyb**
*Based on #137.*

Implements, i.e. closes #128, and uses "anonymous" records for field substructure.

### CODE REVIEW

File: src/generate/templates/header.rs [hunk 1]
```diff
@@ -225,3 +241,54 @@ impl<'a, 'i, I: gll::grammer::input::Input, T> Handle<'a, 'i, I, [T]> {
...
+    pub fn all(self) -> _forest::typed::ShapedAllIter<'a, 'i, _G, I, T> {
+        // HACK(eddyb) this can probably still be wrong sometimes.
+        let mut node = self.node;
+        if let NodeShape::Alias(_) = self.forest.grammar.node_shape(node.kind) {
+            node = self.forest.unpack_alias(node);
+        }
+
+        T::all(self.forest, node)
+    }
+}
```
Comment (CAD97) [inline src/generate/templates/header.rs]: I agree that `Handle` is generally useful enough to live in `grammer`.

File: src/generate/rust.rs [hunk 2]
```diff
@@ -382,10 +487,71 @@ impl<Pat: MatchesEmpty + RustInputPat> GrammarGenerateMethods<Pat> for grammer::
...
         for (&name, &rule) in rules.named {
-            out += declare_rule(name, rule, cx, &mut rules) + impl_parse_with(cx, name);
+            let rust_adt = rule.rust_adt(cx, &mut records);
+            out += declare_rule(name, rule, &rust_adt, cx, &mut rules) + impl_parse_with(cx, name);
         }
 
+        let records = records.into_iter().map(|fields| {
+            let ident = Src::ident(&(fields.join("__") + "__"));
+            // FIXME(eddyb) figure out a more efficient way to reuse
+            // iterators with `quote!(...)` than `.collect::<Vec<_>>()`.
```
Comment (CAD97) [inline src/generate/rust.rs]: What version of `quote` are we using? IIRC `quote:1` behaves better with iterators now. At the very least we can `let f_ident = fields.iter().map(Src::ident); let f_ident_again = f_ident.clone();`.

Comment (eddyb) [reply src/generate/rust.rs]: `proc-quote` and I'd rather not make things *that* ugly.

**Comment by eddyb** [comment]
Maybe I should update the JSON example in the README to not have a separate `Field` rule.

**Pull request merged**